Dockerfile

```diff
@@ -15,71 +18,84 @@ RUN apt-get update \
     apt-get install --assume-yes \
     bash \
     git \
+    rsync \
     python3 \
     postgresql \
     default-jdk \
     openjdk-11-jre \
     build-essential \
     ca-certificates
 
-# switch to user
+# setup folder
 ENV DEST_FOLDER=/home/unlock
-
 RUN mkdir $DEST_FOLDER
+WORKDIR ${DEST_FOLDER}
 
-# copy files 
-WORKDIR /tmp
-COPY . .
-
-# copy all package.json files
-RUN find . -maxdepth 3 -mindepth 2 -type f -name "package.json" ! -path '**node_modules**' ! -path './docker' -exec cp --parents '{}' $DEST_FOLDER \;
-COPY package.json $DEST_FOLDER
-COPY yarn.lock $DEST_FOLDER
-
-# copy al yarn related files
-COPY .yarnrc.yml $DEST_FOLDER
-COPY .prettierrc $DEST_FOLDER
-COPY .yarn/plugins $DEST_FOLDER/.yarn/plugins
-COPY .yarn/releases $DEST_FOLDER/.yarn/releases
-RUN chown -R node:node $DEST_FOLDER
-
-# specify yarn cache folder location (to be used by docker buildkit)
-USER node
-RUN echo "cacheFolder: ${DEST_FOLDER}/yarn-cache" >> $DEST_FOLDER/.yarnrc.yml
-RUN mkdir $DEST_FOLDER/yarn-cache
+# We use rsync to prepare all package.json files that are necessary 
+# for packages install and used to invalidate buidkit docker cache
+RUN --mount=type=bind,target=/docker-context \
+    rsync -amv --delete \
+          --exclude='node_modules' \
+          --exclude='*/node_modules' \
+          --include='package.json' \
+          --include='*/' --exclude='*' \
+          /docker-context/ $DEST_FOLDER;
+
+# yarn related files
+COPY yarn.lock .
+COPY .yarnrc.yml .
+COPY .yarn/plugins .yarn/plugins
+COPY .yarn/releases .yarn/releases
+
+# use custom .yarn/cache folder (for local dev)
+RUN echo "cacheFolder: ${DEST_FOLDER}/yarn-cache" >> .yarnrc.yml
+RUN mkdir ./yarn-cache
+RUN chown -R node:node .
 
 # install deps
-WORKDIR ${DEST_FOLDER}
-RUN --mount=type=cache,target=/home/unlock/yarn-cache,uid=1000,gid=1000 yarn
+USER node
+RUN --mount=type=cache,target=${DEST_FOLDER}/yarn-cache,uid=1000,gid=1000 yarn
 
-#
-# 2. build packages and prepare image for testing/dev
-#
+###################################################################
+# Stage 2: Build packages and app
+###################################################################
 FROM deps as dev
 
-# enforce perms
-WORKDIR /home/unlock
+ENV DEST_FOLDER=/home/unlock
 
-# default user
+# enforce perms
 USER node
 COPY --chown=node . .
 
 # make sure java is installed properly
 RUN java -version
 RUN javac -version
 
+# copy all package.json again
+RUN --mount=type=bind,target=/docker-context \
```

Comment:
Any reason we copy then again? Arguably since we copied all the files again (line 68), they are in there?